### PR TITLE
Chat plugin capability

### PR DESCRIPTION
**Pull request by Wladastic**
### Background

As I am currently building the telegram chat plugin, the plugin capabilities for logs, userinput etc were missing.

### Changes

anywhere the user gets feedback and get asked for input, the input is sent to the according plugin method

### Documentation

can_handle_report and can_handle_user_input needed in plugin __init__.py

handle_input(str) is a sync method that is supposed to work similar to the "input()" method, only that it can be used for example for asking the User a question and waiting for the answer.

report(str)  just sends the user a message like a log or in my case with telegram chat it will just forward all messages that would normally only be visible in the terminal.

To enable, set _CHAT_MESSAGES_ENABLED=True_ in your .env file

### Test Plan

I will finish my telegram chat plugin and then set the missing ticks.
Feel free to also test then, as this is supposed to be available to every type of plugin that needs those methods.

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

**Comment by Wladastic** [comment]
@BillSchumacher You can also give feedback to the naming of the methods. I am not 100% satisfied with "report(str)"
It could also be send_message(str), but not sure if any other plugins are actually being chat plugins then..

File: .env.template
Comment (ntindle) [inline .env.template:209]: Why would this be gated?

Comment (Wladastic) [reply .env.template:209]: I figured it's easier to check that way, so that the plugins don't get iterated everytime an interaction takes place...

Comment (Wladastic) [reply .env.template:209]: Also another thought:
You wouldn't want to keep deleting the plugins just to not use it for once.
Maybe a Plugin Manage would be nice

File: autogpt/auto_gpt_workspace/instance_history.txt
Comment (richbeales) [inline autogpt/auto_gpt_workspace/instance_history.txt]: Should this be here?

Comment (Wladastic) [reply autogpt/auto_gpt_workspace/instance_history.txt]: uhm, no idea how that happened, one second...

File: autogpt/utils.py
Comment (BillSchumacher) [inline autogpt/utils.py]: This potentially errors without the template package being updated, let's check that this attribute exists on the class before calling it.

**Review by BillSchumacher — CHANGES_REQUESTED** [review]

File: autogpt/utils.py
Comment (BillSchumacher) [reply autogpt/utils.py]: the other reports function needs this as well if it has been defined here.

**Review by BillSchumacher — COMMENTED** [review]
LGTM needs manual testing.

**Comment by ntindle** [comment]
@desojo for context

File: autogpt/utils.py
Comment (Pwuts) [inline autogpt/utils.py]: redundant statement; `not ""` evaluates to `true`

File: autogpt/utils.py
Comment (Pwuts) [inline autogpt/utils.py]: redundant after the previous `if` statement

File: autogpt/utils.py
Comment (Pwuts) [inline autogpt/utils.py]:
```suggestion
def send_chat_message_to_user(report: str):
```

File: autogpt/agent/agent.py
Comment (Pwuts) [inline autogpt/agent/agent.py:151]:
```suggestion
                        console_input = clean_input("Waiting for your response via chat plugin...")
```

or something along those lines

File: autogpt/agent/agent.py
Comment (Pwuts) [inline autogpt/agent/agent.py]: A call to `clean_chat_message_to_user` is always nested inside `if cfg.chat_messages_enabled`, so better to just integrate that check in the function.

File: autogpt/config/config.py
Comment (Pwuts) [inline autogpt/config/config.py]:
```suggestion
        self.chat_messages_enabled = os.getenv("CHAT_MESSAGES_ENABLED") == "True"
```

File: autogpt/utils.py
Comment (Pwuts) [inline autogpt/utils.py]: I prefer to have this kind of extensive parsing/mapping in the plugin instead of in the core. Otherwise, reduce the list to the obvious options, and make sure to fix the casing of the `plugin_response` before matching.

**Review by Pwuts — DISMISSED** [review]

File: autogpt/utils.py
Comment (Wladastic) [reply autogpt/utils.py]: Did I make an error with None?
I thought if None or "" should make it continue to the next plugin

File: autogpt/agent/agent.py
Comment (Wladastic) [reply autogpt/agent/agent.py:151]: That would result in it sending you the message "Waiting for your response via chat plugin..." within Telegram for example.
Maybe it should be "What shall I do?" then?

File: autogpt/agent/agent.py
Comment (Wladastic) [reply autogpt/agent/agent.py]: good idea!

File: autogpt/utils.py
Comment (Wladastic) [reply autogpt/utils.py]: I think I will first reduce this then

File: autogpt/utils.py
Comment (Wladastic) [reply autogpt/utils.py]: added the hasattr() method as discussed :)

File: autogpt/utils.py
Comment (Wladastic) [reply autogpt/utils.py]: oh sorry, now I get it.

File: autogpt/agent/agent.py
Comment (Pwuts) [reply autogpt/agent/agent.py:151]: Oh nevermind, I misread.

File: autogpt/utils.py
Comment (Pwuts) [inline autogpt/utils.py]: this is still case sensitive

File: autogpt/utils.py
Comment (Pwuts) [inline autogpt/utils.py]: A less indentation-heavy version of this would be:
```suggestion
    if not cfg.chat_messages_enabled:
        return

    for plugin in enumerate(cfg.plugins):
```

Just a point of feedback, doesn't necessarily need fixing.

Comment (Wladastic) [reply autogpt/utils.py]: accepted, I agree

File: autogpt/utils.py
Comment (Wladastic) [reply autogpt/utils.py]: I put the not plugin_response on top for that, to make sure it's not empty.

File: autogpt/utils.py
Comment (Pwuts) [reply autogpt/utils.py]: Well in that case, https://github.com/Significant-Gravitas/Auto-GPT/pull/2929/files#diff-680674214fd6efa66f980e3ba510374dc0adb02690fd4e642c637bcf669332ffR32-R33 has the same structure :p

Comment (Wladastic) [reply autogpt/utils.py]: not really, there the normal input is the fallback and would also be skipped.

**Comment by Wladastic** [comment]
I had to change enumerate to make the plugin work.
Otherwise it was turned into a tuple which didn't work.

**Comment by Wladastic** [comment]
proof of work:
https://cdn.discordapp.com/attachments/1100128423541747722/1100509590606590112/RPReplay_Final1682452290.mov

File: autogpt/utils.py
Comment (BillSchumacher) [inline autogpt/utils.py]: Let's pass the plugins by value to the functions instead of keeping this global singleton around.

**Review by BillSchumacher — CHANGES_REQUESTED** [review]

File: autogpt/utils.py
Comment (Wladastic) [reply autogpt/utils.py]: How?

Comment (Wladastic) [reply autogpt/utils.py]: You mean as method properties?

Comment (ntindle) [reply autogpt/utils.py]: https://github.com/Significant-Gravitas/Auto-GPT/blob/master/autogpt/agent/agent.py

Comment (ntindle) [reply autogpt/utils.py]: @collijk may also have thoughts about this location

Comment (Wladastic) [reply autogpt/utils.py]: Ah now I get it, one second...

Comment (Wladastic) [reply autogpt/utils.py]: done

**Comment by collijk** [comment]
I have no concerns about the config usage after the update at @BillSchumacher 's request.  Will leave it to you guys to work through any remaining concerns.

**Comment by Wladastic** [comment]
Requested Changes are done! :)

**Review by richbeales — APPROVED** [review]

**Review by ntindle — APPROVED** [review]

**Pull request merged**
